File: Scala/src/main/scala/iterator/Waitress.scala
package iterator

class Waitress(menus: List[Menu]) {

  def printMenu(): Unit = {
    println("메뉴")
    menus.foreach(menu => {
      println("--------")
      println(menu.name)
      printMenu(menu.createUtilIterator)
    })
  }

  def printMenu(iterator: Iterator): Unit = {
    while (iterator.hasNext) {
      val menuItem = iterator.next
      println(menuItem.name + ", " + menuItem.price + " -- " + menuItem.description)
    }
  }

  def printMenu(iterator: java.util.Iterator[MenuItem]): Unit = {
    while (iterator.hasNext) {
      val menuItem = iterator.next
      println(menuItem.name + ", " + menuItem.price + " -- " + menuItem.description)
    }
  }
}

File: Kotlin/src/main/kotlin/chapter9_iterator/Waitress.kt
package chapter9_iterator

class Waitress(
    val pancakeHouseMenu: PancakeHouseMenu,
    val dinerMenu: DinerMenu,
) {
    fun printMenu() {
        val pancakeIterator = pancakeHouseMenu.getIterator()
        val dinerIterator = dinerMenu.createIterator()

        println("#### 아침 메뉴 ####")
        printMenu(pancakeIterator)

        println("#### 점심 메뉴 ####")
        printMenu(dinerIterator)
    }

    fun printMenu(iterator: Iterator) {
        while (iterator.hasNext()) {
            println(iterator.next())
        }
    }
}